Fedora 28 compilation warnings

Compiling with GCC 8 on Fedora 28 (4.16.5-300.fc28.x86_64) produces warnings about string truncation and stack validation. 
```
make -C /lib/modules/4.16.5-300.fc28.x86_64/build M=/root/dattobd/src modules
make[1]: Entering directory '/usr/src/kernels/4.16.5-300.fc28.x86_64'
  CC [M]  /root/dattobd/src/dattobd.o
In file included from ./include/linux/bitmap.h:9,
                 from ./include/linux/cpumask.h:12,
                 from ./arch/x86/include/asm/cpumask.h:5,
                 from ./arch/x86/include/asm/msr.h:11,
                 from ./arch/x86/include/asm/processor.h:21,
                 from ./arch/x86/include/asm/cpufeature.h:5,
                 from ./arch/x86/include/asm/thread_info.h:53,
                 from ./include/linux/thread_info.h:38,
                 from ./arch/x86/include/asm/preempt.h:7,
                 from ./include/linux/preempt.h:81,
                 from ./include/linux/spinlock.h:51,
                 from ./include/linux/seqlock.h:36,
                 from ./include/linux/time.h:6,
                 from ./include/linux/stat.h:19,
                 from ./include/linux/module.h:10,
                 from /root/dattobd/src/includes.h:14,
                 from /root/dattobd/src/dattobd.c:11:
In function ‘strncpy’,
    inlined from ‘string_copy’ at /root/dattobd/src/dattobd.c:3091:2:
./include/linux/string.h:246:9: warning: ‘__builtin_strncpy’ output truncated before terminating nul copying as many bytes from a string as its length [-Wstringop-truncation]
  return __builtin_strncpy(p, q, size);
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/root/dattobd/src/dattobd.c: In function ‘string_copy’:
./include/linux/string.h:267:10: note: length computed here
   return __builtin_strlen(p);
          ^~~~~~~~~~~~~~~~~~~
In function ‘strncpy’,
    inlined from ‘pathname_concat’ at /root/dattobd/src/dattobd.c:1378:2,
    inlined from ‘user_mount_pathname_concat’ at /root/dattobd/src/dattobd.c:1391:8:
./include/linux/string.h:246:9: warning: ‘__builtin_strncpy’ output truncated before terminating nul copying as many bytes from a string as its length [-Wstringop-truncation]
  return __builtin_strncpy(p, q, size);
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/root/dattobd/src/dattobd.c: In function ‘user_mount_pathname_concat’:
./include/linux/string.h:267:10: note: length computed here
   return __builtin_strlen(p);
          ^~~~~~~~~~~~~~~~~~~
In function ‘strncpy’,
    inlined from ‘tracer_dattobd_info’ at /root/dattobd/src/dattobd.c:3925:2,
    inlined from ‘ioctl_dattobd_info’ at /root/dattobd/src/dattobd.c:4188:2,
    inlined from ‘ctrl_ioctl’ at /root/dattobd/src/dattobd.c:4294:9:
./include/linux/string.h:246:9: warning: ‘__builtin_strncpy’ specified bound 4096 equals destination size [-Wstringop-truncation]
  return __builtin_strncpy(p, q, size);
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In function ‘strncpy’,
    inlined from ‘tracer_dattobd_info’ at /root/dattobd/src/dattobd.c:3926:2,
    inlined from ‘ioctl_dattobd_info’ at /root/dattobd/src/dattobd.c:4188:2,
    inlined from ‘ctrl_ioctl’ at /root/dattobd/src/dattobd.c:4294:9:
./include/linux/string.h:246:9: warning: ‘__builtin_strncpy’ specified bound 4096 equals destination size [-Wstringop-truncation]
  return __builtin_strncpy(p, q, size);
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/root/dattobd/src/dattobd.o: warning: objtool: file_io()+0x80: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: __cow_sync_and_free_sections()+0x7e: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: __cow_cleanup_mappings()+0x82: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: __cow_write_header()+0x8c: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: __cow_open_header()+0x4d: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: file_open()+0x21: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: __tracer_destroy_snap()+0x98: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: __tracer_setup_snap()+0x1e: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: __tracer_transition_tracing()+0x82: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: __tracer_setup_tracing()+0x32: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: __tracer_destroy_tracing()+0x1c: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: string_copy()+0x2a: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: path_get_absolute_pathname()+0x2b: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: __tracer_destroy_cow_path.isra.29()+0x19: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: copy_string_from_user.part.32()+0x1a: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: get_reload_params()+0x3f: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: user_mount_pathname_concat()+0x9a: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: __tracer_destroy_base_dev()+0x4e: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: file_get_absolute_pathname.isra.36()+0x3e: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: on_bio_read_complete()+0x22: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: inc_make_sset()+0x2d: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: __tracer_destroy_cow_thread.isra.44()+0x19: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: __tracer_setup_cow_thread()+0x1b: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: __tracer_setup_base_dev()+0x2d: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: __file_unlink.constprop.55()+0x53: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: __tracer_destroy_cow()+0x3c: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: __tracer_setup_unverified()+0x3c: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: __tracer_setup_cow_path.isra.40()+0x1c: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: snap_open()+0x17: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: snap_release()+0x10: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: __cow_load_section()+0x3b: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: __cow_write_mapping()+0xcb: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: inc_sset_thread()+0x102: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: cow_read_mapping()+0xab: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: tracer_alloc()+0x12: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: snap_handle_read_bio()+0x36d: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: snap_cow_thread()+0x526: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: tracing_mrf()+0x5ad: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: __tracer_setup_cow()+0x3f8: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: __ioctl_setup()+0x46: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: auto_transition_active()+0x2ad: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: handle_bdev_mount_event()+0x1c8: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: post_umount_check.part.49()+0x7c: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: umount_hook()+0x2a: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: mount_hook()+0x4f: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: ioctl_transition_inc()+0x2d: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: ctrl_ioctl()+0x60: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: agent_exit()+0x13: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: on_bio_read_complete.cold.77()+0x45: sibling call from callable instruction with modified stack frame
/root/dattobd/src/dattobd.o: warning: objtool: snap_release.cold.87()+0xd: sibling call from callable instruction with modified stack frame
  Building modules, stage 2.
  MODPOST 1 modules
  CC      /root/dattobd/src/dattobd.mod.o
  LD [M]  /root/dattobd/src/dattobd.ko
make[1]: Leaving directory '/usr/src/kernels/4.16.5-300.fc28.x86_64'

```

--- a/src/dattobd.c
+++ b/src/dattobd.c
@@ -4666,12 +4666,15 @@ static void post_umount_check(int dormant_ret, long umount_ret, unsigned int idx
 	task_work_flush();
 
 	//if we went dormant, but the block device is still mounted somewhere, goto fail state
-	sb = get_super(dev->sd_base_dev);
+	sb = dattobd_get_super(dev->sd_base_dev);
 	if(sb){
-		LOG_ERROR(-EIO, "device still mounted after umounting cow file's file-system. entering error state");
-		tracer_set_fail_state(dev, -EIO);
-		drop_super(sb);
-		return;
+		if(!(sb->s_flags & MS_RDONLY)){
+			LOG_ERROR(-EIO, "device still mounted after umounting cow file's file-system. entering error state");
+			tracer_set_fail_state(dev, -EIO);
+			dattobd_drop_super(sb);
+			return;
+		}
+		dattobd_drop_super(sb);
 	}
 
 	LOG_DEBUG("post umount check succeeded");

--- a/src/dattobd.c
+++ b/src/dattobd.c
@@ -3076,32 +3076,6 @@ static int snap_merge_bvec(struct request_queue *q, struct bio *bio_bvm, struct 
 
 /*******************************SETUP HELPER FUNCTIONS********************************/
 
-static int string_copy(char *str, char **dest){
-	int ret, len;
-	char *out;
-
-	len = strlen(str);
-	out = kmalloc(len + 1, GFP_KERNEL);
-	if(!out){
-		ret = -ENOMEM;
-		LOG_ERROR(ret, "error allocating destination for string copy");
-		goto error;
-	}
-
-	strncpy(out, str, len);
-	out[len] = '\0';
-
-	*dest = out;
-	return 0;
-
-error:
-	LOG_ERROR(ret, "error copying string");
-	if(out) kfree(out);
-
-	*dest = NULL;
-	return ret;
-}
-
 static int bdev_is_already_traced(struct block_device *bdev){
 	int i;
 	struct snap_device *dev;
@@ -3740,8 +3714,6 @@ error:
 }
 
 static int __tracer_setup_unverified(struct snap_device *dev, unsigned int minor, char *bdev_path, char *cow_path, unsigned long cache_size, int is_snap){
-	int ret;
-
 	if(is_snap) set_bit(SNAPSHOT, &dev->sd_state);
 	else clear_bit(SNAPSHOT, &dev->sd_state);
 	clear_bit(ACTIVE, &dev->sd_state);
@@ -3749,13 +3721,11 @@ static int __tracer_setup_unverified(struct snap_device *dev, unsigned int minor
 
 	dev->sd_cache_size = cache_size;
 
-	//copy the bdev_path
-	ret = string_copy(bdev_path, &dev->sd_bdev_path);
-	if(ret) goto error;
+	dev->sd_bdev_path = kstrdup(bdev_path, GFP_KERNEL);
+	if(!dev->sd_bdev_path) goto error;
 
-	//copy the cow_path
-	ret = string_copy(cow_path, &dev->sd_cow_path);
-	if(ret) goto error;
+	dev->sd_cow_path = kstrdup(cow_path, GFP_KERNEL);
+	if(!dev->sd_cow_path) goto error;
 
 	//add the tracer to the array of devices
 	__tracer_setup_tracing_unverified(dev, minor);
@@ -3763,9 +3733,9 @@ static int __tracer_setup_unverified(struct snap_device *dev, unsigned int minor
 	return 0;
 
 error:
-	LOG_ERROR(ret, "error setting up unverified tracer");
+	LOG_ERROR(-ENOMEM, "error setting up unverified tracer");
 	tracer_destroy(dev);
-	return ret;
+	return -ENOMEM;
 }
 #define tracer_setup_unverified_inc(dev, minor, bdev_path, cow_path, cache_size) __tracer_setup_unverified(dev, minor, bdev_path, cow_path, cache_size, 0)
 #define tracer_setup_unverified_snap(dev, minor, bdev_path, cow_path, cache_size) __tracer_setup_unverified(dev, minor, bdev_path, cow_path, cache_size, 1)
@@ -3922,8 +3892,8 @@ static void tracer_dattobd_info(struct snap_device *dev, struct dattobd_info *in
 	info->state = dev->sd_state;
 	info->error = tracer_read_fail_state(dev);
 	info->cache_size = (dev->sd_cache_size)? dev->sd_cache_size : dattobd_cow_max_memory_default;
-	strncpy(info->cow, dev->sd_cow_path, PATH_MAX);
-	strncpy(info->bdev, dev->sd_bdev_path, PATH_MAX);
+	strlcpy(info->cow, dev->sd_cow_path, PATH_MAX);
+	strlcpy(info->bdev, dev->sd_bdev_path, PATH_MAX);
 
 	if(!test_bit(UNVERIFIED, &dev->sd_state)){
 		info->falloc_size = dev->sd_cow->file_max;

